### PR TITLE
Remove `Hosts` and `File` bootstrapOptions

### PR DESCRIPTION
**Pull request by Motiejus**
They can be replaced by implementing a DiscoverProvider instead. We
provide two default implementations:

1. discovery.statichosts (compatible with old `Hosts` attribute).
2. discovery.staticjson (compatible with old `File` attribute).

A separate pull request which fixes `Bootstrap()` not to have the `StaticHosts`-specific workaround will follow.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -1,6 +1,12 @@
 ringpop-go changes
 ==================
 
+v0.x.y (to be released)
```
Comment (thanodnl) [inline CHANGES.md]: Love the addition of this to the change log. Was thinking of exactly this so we keep track during development on what gets added. However I was thinking about calling this `Dev` since we don't know the version to be released. What are you thoughts on calling this section something like `Dev` or `Development`?

File: discovery/staticjson/lib.go [hunk 2]
```diff
@@ -18,11 +18,10 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
-package swim
+package staticjson
```
Comment (thanodnl) [inline discovery/staticjson/lib.go]: not really sure if this package name captures what it does.

1. It is not necessarily static as it reads the file every time you call `Hosts`
2. This name does not tell that it is reading the json from a file

I would suggest `jsonfile` as the package name.

File: ringpop.go [hunk 3]
```diff
@@ -330,11 +331,10 @@ func (rp *Ringpop) setState(s state) {
 // Bootstrap starts communication for this Ringpop instance.
 //
 // When Bootstrap is called, this Ringpop instance will attempt to contact
-// other instances from a seed list provided either in the BootstrapOptions or
-// as a JSON file.
+// other instances from the DiscoverProvider.
 //
 // If no seed hosts are provided, a single-node cluster will be created.
-func (rp *Ringpop) Bootstrap(userBootstrapOpts *swim.BootstrapOptions) ([]string, error) {
+func (rp *Ringpop) Bootstrap(userbootstrapOpts *swim.BootstrapOptions) ([]string, error) {
```
Comment (thanodnl) [inline ringpop.go]: Is there a specific reason that you changed that capitalization of `userBootstrapOpts` ? I find the one with capital `B` easier to read.

File: ringpop.go [hunk 4]
```diff
@@ -346,14 +346,23 @@ func (rp *Ringpop) Bootstrap(userBootstrapOpts *swim.BootstrapOptions) ([]string
 	if err != nil {
 		return nil, err
 	}
+	hosts, err := userbootstrapOpts.DiscoverProvider.Hosts()
+	if err != nil {
+		rp.logger.WithField("error", err).Info("Failed to get boostrap hosts")
```
Comment (thanodnl) [inline ringpop.go]: You are logging an `error` on `info` level this seems counterintuitive.

File: ringpop.go [hunk 5]
```diff
@@ -346,14 +346,23 @@ func (rp *Ringpop) Bootstrap(userBootstrapOpts *swim.BootstrapOptions) ([]string
...
-	// check we're in the bootstrap host list and add ourselves if we're not
+	// Check we're in the bootstrap host list and add ourselves if we're not
 	// there. If the host list is empty, this will create a single-node
 	// cluster.
-	bootstrapOpts := *userBootstrapOpts
-	if len(bootstrapOpts.File) == 0 && !util.StringInSlice(bootstrapOpts.Hosts, identity) {
-		bootstrapOpts.Hosts = append(bootstrapOpts.Hosts, identity)
+	// TODO: move this code to the caller of the Hosts().
+	bootstrapOpts := *userbootstrapOpts
+	switch bootstrapOpts.DiscoverProvider.(type) {
```
Comment (thanodnl) [inline ringpop.go]: I thought we discussed to treat all the `DiscoveryProviders` the same. What happens if there is an other provider that does not report the current node as part of the list?

File: discovery/statichosts/lib.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+// Hosts just returns the static list of hosts that was provided to this struct
+// on construction.
+func (p *StaticHostList) Hosts() ([]string, error) {
+	return p.hosts, nil
+}
+
+// New creates a provider with a static list of hosts.
+func New(hosts []string) *StaticHostList {
```
Comment (thanodnl) [inline discovery/statichosts/lib.go]: Have you considered using [`hosts ...string`](https://golang.org/ref/spec#Passing_arguments_to_..._parameters) here? It makes it easier to use if you hardcode the hosts in your application since you don't have to initialize the slice your self.

If you happen to already have a slice you can still pass it in to this function via `New(yourHostsLists...)`

File: CHANGES.md (hunk 1 above)
Comment (dansimau) [inline CHANGES.md]: I also did this, see:
https://github.com/uber/ringpop-go/pull/118/files#diff-8b1c3fd0d4a6765c16dfd18509182f9dR4

I called it "0.3.1", as I intend it to be in the next _minor_ release.

File: discovery/staticjson/lib.go (hunk 2 above)
Comment (dansimau) [inline discovery/staticjson/lib.go]: +1 with Niels. I thought the same thing.

File: ringpop.go (hunk 3 above)
Comment (dansimau) [inline ringpop.go]: +1 Niels. Why was this changed?

File: discovery/staticjson/lib.go (hunk 2 above)
Comment (thanodnl) [inline discovery/staticjson/lib.go]: **UCH** Nils **UCH**

File: CHANGES.md (hunk 1 above)
Comment (severb) [inline CHANGES.md]: I think whatever works as long as we don't have a fixed number. I used "unreleased" before.

Comment (Motiejus) [inline CHANGES.md]: I like the idea. `0.DEV`.

File: discovery/staticjson/lib.go (hunk 2 above)
Comment (Motiejus) [inline discovery/staticjson/lib.go]: OK.

File: ringpop.go (hunk 3 above)
Comment (Motiejus) [inline ringpop.go]: I refactored the whole section, `userbootstrapOpts` and the likes are gone from this file.

Comment (dansimau) [inline ringpop.go]: I think the point is "userbootstrap" is not a word. So if you're doing camel case, it should be "userBootstrap"...

File: ringpop.go (hunk 4 above)
Comment (Motiejus) [inline ringpop.go]: Removed this message completely in the other incarnation.

File: ringpop.go (hunk 5 above)
Comment (Motiejus) [inline ringpop.go]: That was the idea to first post a compatible change, and later, in a separate PR, fix that (see the PR message).

I will send an update with this section fixed.

File: discovery/statichosts/lib.go (hunk 6 above)
Comment (Motiejus) [inline discovery/statichosts/lib.go]: Great idea.

**Comment by Motiejus** [comment]
Like discussed with @thanodnl, I`ll add some explicit tests for DiscoverProviders.

File: CHANGES.md [hunk 7]
```diff
@@ -1,6 +1,14 @@
 ringpop-go changes
 ==================
 
+v0.DEV (to be released)
+-----------------------
+
+* Remove `File` and `Host`-based discover providers in favor of
```
Comment (dansimau) [inline CHANGES.md]: This is a breaking change, so I think we need to provide an additional section: "release notes" which gives clear instructions on how to migrate from the old API to the new. There is a "release notes" section further down in this readme to get an example from.

Comment (Motiejus) [inline CHANGES.md]: Good one, thanks! Added.

**Comment by thanodnl** [comment]
LGTM

**Comment by dansimau** [comment]
LGTM. 

I think we could do some bike shedding over the exact API, e.g. here's another option:

```
import github.com/uber/ringpop-go/discovery

discovery.StaticHostList([]string{"1.1.1.1:1234", "1.1.1.2:1234"})
discovery.JSONFile("/path/to/file")
```

I think this is a nicer/cleaner API, though, I am undecided about whether it's as idiomatic in Go (or whether that even matters!). I can more imagine this type of API from Python (a la Kenneth's "API For Humans" style).

What do you guys think?

**Comment by Motiejus** [comment]
@dansimau While I agree the API would look less to write and potentially prettier, we agreed to put every discovery provider to its own separate package, rather than have a single package for the discovery providers. Merging.

**Comment by Motiejus** [comment]
Squashed all the commits and rebased on top of dev. Will wait for the final round of integ tests, and do a fast-forward merge.
